### ISSUE TITLE
ERC2771Forwarder Type Hash Inconsistency

### ISSUE THREAD

**Issue by JaredBorders** [OpenZeppelin/openzeppelin-contracts#4698]
## Summary 

`ERC2771Forwarder.ForwardRequestData` struct does not match `_FORWARD_REQUEST_TYPEHASH`

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/149e1b79fecb9db32f732c8c9f05ac2d8fa97471/contracts/metatx/ERC2771Forwarder.sol#L65

## 💻 Environment

v5.0.0

## 📝 Details

The struct is defined as:

```
struct ForwardRequestData {
        address from;
        address to;
        uint256 value;
        uint256 gas;
        uint48 deadline;
        bytes data;
        bytes signature;
}
```

and the type hash is defined as:

```
bytes32 internal constant _FORWARD_REQUEST_TYPEHASH =
    keccak256(
        "ForwardRequest(address from,address to,uint256 value,uint256 gas,uint256 nonce,uint48 deadline,bytes data)"
    );
```

The type hash parameters should match the fields found in the struct.

## 🔢 Code to reproduce bug

N/A

**Comment by ernestognw** [comment]
Hey @JaredBorders, thanks for reporting. However, we don't see this mismatch as an issue. The `signature` value is not included because you'd require the signature to sign a message. And regarding the `nonce`, we discussed it during the 5.0 development and reached the following conclusions:

1.  Including `nonce` in `ForwardRequestData` is redundant information because the nonce is a 1:1 relationship with the address.
2. Removing `nonce` from `ForwardRequestData` reduced the calldata size, which is especially important for the `executeBatch` function.
3. Given meta transactions are sponsored by an entity, it's in its best interest to save in calldata costs.
4. The nonce can't be spoofed anyway because it needs to be included in the signed message (what the end user will see in metamask)

> The type hash parameters should match the fields found in the struct.

Can you elaborate on why they should? I'd be happy to hear if we're missing something with the design of the forwarder

**Comment by JaredBorders** [comment]
>  The signature value is not included because you'd require the signature to sign a message

Okay, this makes sense to me. 

> Can you elaborate on why they should? I'd be happy to hear if we're missing something with the design of the forwarder

I re-read EIP-712 and quoting from it: 

"Definition: A struct type has valid identifier as name and contains zero or more member variables. Member variables have a member type and a name." 

This does *seem* to support the concept of not including all member variables of a struct in its typehash. Leaving members out is uncommon in my experience, though, which is why I opened the issue. 

> conclusions 1 - 4

These make sense to me. However, including a nonce would allow for improved signer control over requests. 

For example, if a nonce was included, it could be invalidated by the user *prior to execution*. Invalidation would require more logic though, and would subsequently increase the gas required to check that a *specific* nonce hasn't been used.

This is likely a trade-off that has already been considered, though. 

-

I will wait for your response, but I think this issue can be closed. Thank you for your swift reply @ernestognw 🫡

**Comment by ernestognw** [comment]
> This does seem to support the concept of not including all member variables of a struct in its typehash. Leaving members out is uncommon in my experience, though, which is why I opened the issue.

Yeah, also note that the definition also specifies that is not Solidity-specific and _aims to be agnostic to higher-level languages_. Not that our library is a different language but the standard allows us to play a bit with its definition, in reality, the definition of `𝕊` is an implicit subset of the `ForwardRequestData` struct.

We also named the struct `ForwardRequestData` to clearly differentiate it from the `𝕊` identifier, which is `ForwardRequest` (included in the `_FORWARD_REQUEST_TYPEHASH` encoding).

> For example, if a nonce was included, it could be invalidated by the user prior to execution. Invalidation would require more logic though, and would subsequently increase the gas required to check that a specific nonce hasn't been used.

Remember the [nonce is included in `𝕊` encoding](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/3e6c86392c97fbc30d3d20a378a6f58beba08eba/contracts/metatx/ERC2771Forwarder.sol#L229) but not in calldata. The user can still invalidate a transaction by relaying another transaction with the same nonce. The `Nonces` contract guarantees nonces can't be used more than once and it's [formally verified](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/3eb5cfb22a589a84805dc631cece72c5c10a35fa/certora/specs/Nonces.spec).

> Thank you for your swift reply @ernestognw 🫡

It's a pleasure friend 🤓. Closing
